### PR TITLE
Polish fleet API to support cuda collective mode and nccl2 mode.

### PR DESCRIPTION
**Pull request by gongweibao**
Polish fleet API to support cuda collective mode and nccl2 mode.
- Polish DistributedStrategy members and support dist_fc and local_sgd
- Polish minimize function to support compiler function.

### CODE REVIEW

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 1]
```diff
@@ -260,24 +286,20 @@ def minimize(self,
...
+        self.startup_program = startup_program
 
-        worker_endpoints = fleet.worker_endpoints()
-        trainer_id = fleet.worker_index()
-        current_endpoint = fleet.worker_endpoints()[trainer_id]
+        self._loss = loss
 
-        startup_program = startup_program if startup_program else \
-            fluid.framework.default_startup_program
+        _check(main_program, self._optimizer, self._strategy)
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py]: could you specify what kind of check does this function do? maybe a detailed function name is needed

Comment (gongweibao) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py]: Done.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 2]
```diff
@@ -117,23 +78,65 @@ def save_inference_model(self,
...
         io.save_inference_model(dirname, feeded_var_names, target_vars,
-                                self._executor, main_program, None, None,
+                                executor, main_program, None, None,
                                 export_for_deployment)
 
     def save_persistables(self, executor, dirname, main_program=None):
-        io.save_persistables(self._executor, dirname, main_program, None)
+        io.save_persistables(executor, dirname, main_program, None)
+
+    def node_num(self):
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:88]: I think this function is not in fleet_base.py, we have a protocol that all sub class implementation should follow fleet_base.py inferface design.

Comment (gongweibao) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:88]: Done.

Comment (gongweibao) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py]: Done.

**Review by gavin1332 — APPROVED** [review]
LGTM

**Comment by luotao1** [comment]
```
[05:35:17]	[Step 1/1] + echo 'You must have one RD (XiaoguangHu01,chengduoZH,Xreki,luotao1,sneaxiy,tensor-tang) approval for the api change! python/paddle/fluid/framework.py for the management reason of the underlying code for fluid.'
```

**Review by luotao1 — APPROVED** [review]
LGTM for the update on `framework.py`

**Review by sandyhouse — APPROVED** [review]
LGTM

**Pull request merged**